File: AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldControllerTest.java
package com.ichi2.anki.multimediacard.fields;

import android.app.Activity;
import android.content.ContentResolver;
import android.content.Intent;

import com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity;
import com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivityTestBase;
import com.ichi2.testutils.AnkiAssert;
import com.ichi2.testutils.MockContentResolver;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.robolectric.RobolectricTestRunner;

import java.io.File;

import androidx.annotation.CheckResult;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.instanceOf;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.when;


@RunWith(RobolectricTestRunner.class)
public class BasicImageFieldControllerTest extends MultimediaEditFieldActivityTestBase {

    @Test
    public void constructionWithoutDataGivesNoError() {
        IFieldController controller = getValidControllerNoImage();
        assertThat(controller, instanceOf(BasicImageFieldController.class));
    }

    @Test
    public void constructionWithDataSucceeds() {
        grantCameraPermission();

        IFieldController controller = getControllerForField(imageFieldWithData(), getEmptyNote(), 0);

        assertThat(controller, instanceOf(BasicImageFieldController.class));
    }

    @Test
    public void nonExistingFileDoesNotDisplayPreview() {
        BasicImageFieldController controller = getValidControllerNoImage();
        assertThat(controller.isShowingPreview(), is(false));

        File f =  Mockito.mock(File.class);
        when(f.exists()).thenReturn(false);

        controller.setImagePreview(f, 100);

        assertThat("A non existing file should not display a preview",
                controller.isShowingPreview(),
                is(false));
    }

    @Test
    public void erroringFileDoesNotDisplayPreview() {
        BasicImageFieldController controller = getValidControllerNoImage();
        assertThat(controller.isShowingPreview(), is(false));

        File f =  Mockito.mock(File.class);
        when(f.exists()).thenReturn(true); //true, but it'll throw due to being a mock.

        controller.setImagePreview(f, 100);

        assertThat("A broken existing file should not display a preview",
                controller.isShowingPreview(),
                is(false));
    }


    @Test
    public void invalidImageResultDoesNotCrashController() {
        BasicImageFieldController controller = getValidControllerNoImage();
        MultimediaEditFieldActivity activity = setupActivityMock(controller, controller.mActivity);

        ContentResolver mock = MockContentResolver.returningEmptyCursor();
        when(activity.getContentResolver()).thenReturn(mock);

        //Act & Assert
        AnkiAssert.assertDoesNotThrow(() -> performImageResult(controller, new Intent()));
    }


    private void performImageResult(BasicImageFieldController controller, Intent intent) {
        controller.onActivityResult(BasicImageFieldController.ACTIVITY_SELECT_IMAGE, Activity.RESULT_OK, intent);
    }

    @CheckResult
    protected BasicImageFieldController getValidControllerNoImage() {
        grantCameraPermission();

        return (BasicImageFieldController) getControllerForField(emptyImageField(), getEmptyNote(), 0);
    }


    private static IField emptyImageField() {
        return new ImageField();
    }

    private static IField imageFieldWithData() {
        IField field = emptyImageField();
        field.setImagePath("test");
        return field;
    }
}

File: AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldControllerTest.kt
/*
 *  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free Software
 *  Foundation; either version 3 of the License, or (at your option) any later
 *  version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along with
 *  this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.ichi2.anki.multimediacard.fields

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import androidx.activity.result.ActivityResultRegistry
import androidx.activity.result.contract.ActivityResultContract
import androidx.annotation.CheckResult
import androidx.core.app.ActivityOptionsCompat
import com.ichi2.anki.R
import com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivityTestBase
import com.ichi2.testutils.AnkiAssert
import com.ichi2.testutils.MockContentResolver
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.mock
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner
import org.robolectric.shadows.ShadowToast
import java.io.File
import kotlin.test.fail

@RunWith(RobolectricTestRunner::class)
open class BasicImageFieldControllerTest : MultimediaEditFieldActivityTestBase() {
    @Test
    fun constructionWithoutDataGivesNoError() {
        @SuppressLint("CheckResult")
        val controller: IFieldController = validControllerNoImage
        assertThat("construction of image field without data should not give an error", controller, instanceOf(BasicImageFieldController::class.java))
    }

    @Test
    fun constructionWithDataSucceeds() {
        val controller = getControllerForField(imageFieldWithData(), emptyNote, 0)
        assertThat("construction of image field with data should succeed", controller, instanceOf(BasicImageFieldController::class.java))
    }

    @Test
    fun nonExistingFileDoesNotDisplayPreview() {
        @SuppressLint("CheckResult")
        val controller = validControllerNoImage
        assertThat(controller.isShowingPreview, equalTo(false))
        val f = mock(File::class.java)
        whenever(f.exists()).thenReturn(false)
        controller.setImagePreview(f, 100)
        assertThat(
            "A non existing file should not display a preview",
            controller.isShowingPreview,
            equalTo(false)
        )
    }

    @Test
    fun erroringFileDoesNotDisplayPreview() {
        @SuppressLint("CheckResult")
        val controller = validControllerNoImage
        assertThat(controller.isShowingPreview, equalTo(false))
        val f = mock(File::class.java)
        whenever(f.exists()).thenReturn(true) // true, but it'll throw due to being a mock.
        controller.setImagePreview(f, 100)
        assertThat(
            "A broken existing file should not display a preview",
            controller.isShowingPreview,
            equalTo(false)
        )
    }

    @Test
    fun fileSelectedOnSVG() {
        @SuppressLint("CheckResult")
        val controller = validControllerNoImage
        val f = File("test.svg")
        controller.setImagePreview(f, 100)
        assertThat(
            "A SVG image file can't be previewed",
            ShadowToast.getTextOfLatestToast(),
            equalTo(getResourceString(R.string.multimedia_editor_svg_preview))
        )
        assertThat("A SVG image file can't be previewed", controller.isShowingPreview, equalTo(false))
    }

    @Test
    fun invalidImageResultDoesNotCrashController() {
        // TODO: This started failing after API 30:
        //  showThemedToast threw an NPE. Diagnose the underlying issue.

        @SuppressLint("CheckResult")
        val controller = validControllerNoImage
        controller.registryToUse = object : ActivityResultRegistry() {
            override fun <I, O> onLaunch(
                requestCode: Int,
                contract: ActivityResultContract<I, O>,
                input: I,
                options: ActivityOptionsCompat?
            ) {
                fail("Unexpected access to the activity result registry!")
            }
        }
        val activity = setupActivityMock(controller, controller.getActivity())
        val mock = MockContentResolver.returningEmptyCursor()
        whenever(activity.contentResolver).thenReturn(mock)

        // Act & Assert
        AnkiAssert.assertDoesNotThrow { performImageResult(controller, Intent()) }
    }

    private fun performImageResult(controller: BasicImageFieldController, intent: Intent) {
        controller.onActivityResult(BasicImageFieldController.ACTIVITY_SELECT_IMAGE, Activity.RESULT_OK, intent)
    }

    @get:CheckResult
    protected val validControllerNoImage: BasicImageFieldController
        get() {
            return getControllerForField(emptyImageField(), emptyNote, 0) as BasicImageFieldController
        }

    private fun imageFieldWithData(): IField {
        val field = emptyImageField()
        field.imagePath = targetContext.externalCacheDir.toString() + "/temp-photos/test"
        return field
    }

    companion object {
        private fun emptyImageField(): IField {
            return ImageField()
        }
    }
}
